Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/TapToPlace/TapToPlace.cs

```diff
@@ -0,0 +1,467 @@
+                ToggleSpatialMeshVisibility(false);
+            }
+
+            SolverHandler.UpdateSolvers = false;
+
+            IsBeingPlaced = false;
+
+            CoreServices.InputSystem?.UnregisterHandler<IMixedRealityPointerHandler>(this);
+        }
+
+        /// <inheritdoc/>
+        public override void SolverUpdate()
+        {
+            // Make sure the Transform target is not null, added for the case where auto start is true 
+            // and the tracked target type is the controller ray, if the hand is not in the frame we cannot
+            // calculate the postion of the object
+            if (SolverHandler.TransformTarget != null)
+            {
+                PerformRaycast();
+                SetPosition();
+                SetRotation();
+            }
+        }
+
+        protected virtual void PerformRaycast()
+        {
+            // The transform target is the transform of the TrackedTargetType, i.e. Controller Ray, Head or Hand Joint
+            var transform = SolverHandler.TransformTarget;
+
+            Vector3 origin = transform.position;
+            Vector3 endpoint = transform.position + transform.forward;
+            currentRay.UpdateRayStep(ref origin, ref endpoint);
+
+            // Check if the current ray hits a magnetic surface
+            DidHit = MixedRealityRaycaster.RaycastSimplePhysicsStep(currentRay, MaxRaycastDistance, MagneticSurfaces, false, out currentHit);  
+        }
+
+        protected virtual void SetPosition()
+        {
+            // Change the position of the GameObjectToPlace if there was a hit, if not then place the object at the default distance
+            // relative to the TrackedTargetType position
+            
+            if (DidHit)
+            {
+                // take the current hit point and add an offset relative to the surface to avoid half of the object in the surface
+                GoalPosition = currentHit.point;  
+                AddOffset(currentHit.normal * SurfaceNormalOffset);
+                
+                // Draw the normal of the raycast hit for debugging 
+                Debug.DrawRay(currentHit.point, currentHit.normal * 0.5f, Color.yellow);
+            }
+            else
+            {
+                GoalPosition = SolverHandler.TransformTarget.position + (SolverHandler.TransformTarget.forward * DefaultPlacementDistance);
+            }
+        }
+
+        protected virtual void SetRotation()
+        {
+            Vector3 direction = currentRay.Direction;
+            Vector3 surfaceNormal = currentHit.normal;
+
+            if (KeepOrientationVertical)
+            {
+                direction.y = 0;
+                surfaceNormal.y = 0;
+            }
+
+            // if the object is on a surface then change the rotation according to the normal of the hit point
+            if (DidHit && rotateAccordingToSurface)
+            {
+                GoalRotation = Quaternion.LookRotation(-surfaceNormal, Vector3.up);
```

Comment:
Also out of curiosity, what's the experience for both KeepOrientationVertical and RotateAccordingToSurface being enabled at the same time?

Reply:
The blue cube has both `RotateAccordingToSurface` and `KeepOrientationVertical` set to true and the Quad has only `RotateAccordingToSurface` true

![TapToPlaceRotateAccordingToSurface](https://user-images.githubusercontent.com/53493796/73895787-a2cc8380-4835-11ea-95d7-44b59062bb89.gif)